varinfo/variable.py

```diff
@@ -235,6 +235,25 @@ def is_temporal(self) -> bool:
         """
         return ' since ' in self.attributes.get('units', '')
 
+    def is_science(self) -> bool:
+        """ Determine if a variable is a science variable.
+            A science variable is classified as a variable that is not a
+            geographic, temporal, or a projected spatial dimension variable,
```

Comment:
I think this definition is a tiny bit wonky. I was expecting that we'd be iterating through the dimensions on the variable and checking that any of those were geographic, temporal or projected x/y. Something like:

```
if any(
    dimension.is_geographic()
    or dimension.is_temporal()
    or dimension.is_projection_x_or_y()
    for dimension in self.references.get('dimensions', [])
):
    return True
```

Also, I think the second conditional block is kind of the reverse... if a variable has `coordinates` or `grid_mapping`, I would have thought that they _are_ science variables. For example, I'm thinking of a swath science variable for the `coordinates` case.

(That would mean the default return value is `False`)

Reply:
Updating with a summary of Slack conversation...

This isn't as easy as I'd first hoped, because the `VariableBase` class does not have access to the parent `VarInfoBase` instance. (Circular dependencies and whatnot). Right now the values in the `VariableBase.references` dictionary are sets of variable paths, not the `VariableBase` instances associated with those variables.

The alternative, to avoid circular dependencies is to move the `VariableBase.is_science` method to be a `VarInfoBase.is_science_variable` method. (This allows us to access other variables in the granule, so the dimensions can be assessed)

I think this means:

1) `VariableBase.is_science` can become `VarInfoBase.is_science_variable` (disclaimer, there maybe minor typos below...):

```
def _is_spatiotemporal_dimension(self, dimension_path: str) -> bool:
    """ A helper method to simplify the list comprehension in `is_science_variable`. """
    dimension = self.get_variable(dimension_path)
    return (
        dimension is not None
        and (dimension.is_geographic or dimension.is_temporal() or dimension.is_projection_x_or_y())
    )

def is_science_variable(self, variable: VariableBase) -> bool:
    if any(
        self._is_spatiotemporal_dimension(dimension)
        for dimension in variable.references.get('dimensions', [])
    ):
            return True

    if (
        variable.references.get('coordinates') is not None
        or variable.references.get('grid_mapping') is not None
    ):
        return True

    return False
```

2) The `VarInfoBase.get_science_variables` method needs to update its set comprehension to:

```
filtered_with_coordinates = {
    variable_path
    for variable_path, variable
    in self.variables.items()
    if variable_path is not None
    and self.is_science_variable(variable)
    and not self.variable_is_excluded(variable_path,
                                      exclusions_pattern)
}
```

3) The `VarInfoBase.get_metadata_variables` method needs to update its set comprehension to:

```
non_coordinate_variables = {
    variable_path
    for variable_path, variable
    in self.variables.items()
    if variable_path is not None
    and (self.variable_is_excluded(variable_path, exclusions_pattern)
         and not self.is_science_variable(variable))
}
```

(Sorry for the essay!)

Reply:
No sorry's! I needed the essay!!!

Reply:
* [ ] updated

Reply:
And here is where I think that error was introduced. :stare: